### PR TITLE
Migrate `input-utils` to TypeScript

### PR DESCRIPTION
**Pull request by adnasa**
#### Summary

- migrated `input-utils` to TypeScript
- blocked by `buttons` migration , so I set the target of this PR to #1856

### CODE REVIEW

File: packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx [hunk 1]
```diff
@@ -1,31 +1,53 @@
...
-  const handleHeightChange = React.useCallback(
-    (height, meta) => {
-      const rowCount = Math.floor(ref.current.scrollHeight / meta.rowHeight);
-      onHeightChange(height, rowCount);
+  const handleHeightChange = React.useCallback<
+    (height: number, meta?: TextareaHeightChangeMeta) => void
+  >(
+    (height: number, meta?: TextareaHeightChangeMeta) => {
+      // @ts-ignore
+      const rowCount = Math.floor(ref?.current?.scrollHeight / meta?.rowHeight);
```
Comment (adnasa) [inline packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx:37]: I'm still unsure how to address this, so I added an `ignore`
image

File: packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx [hunk 2]
```diff
@@ -1,31 +1,53 @@
...
+      if (onHeightChange) {
+        onHeightChange(height, rowCount);
+      }
     },
     [ref, onHeightChange]
   );
   return (
     <TextareaAutosize
-      ref={ref}
```
Comment (mohib0306) [inline packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx:22]: why is this removed?

**Review by mohib0306 — COMMENTED** [review]
LGTM

File: packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx (hunk 2 above)
Comment (adnasa) [reply packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx:22]: good catch.
This might have to be reverted, although I'll `ts-ignore`  since `ref` is not exposed as `props` on the lib

File: packages/components/inputs/input-utils/src/localized-input-toggle/localized-input-toggle.tsx [hunk 3]
```diff
@@ -1,21 +1,42 @@
...
 
-const LocalizedInputToggle = (props) => {
-  const intl = useIntl();
+type TLocalizedInputToggleProps = {
+  icon?: React.ReactElement;
+  isOpen?: boolean;
+  /**
+   * Handler when the button is clicked
+   */
+  onClick?: (
```
Comment (mohib0306) [inline packages/components/inputs/input-utils/src/localized-input-toggle/localized-input-toggle.tsx:14]: This prop is required in the old code?

**Review by mohib0306 — APPROVED** [review]
LGTM. Is the readme regenerated (up-to-date) after the prop definition?

**Comment by adnasa** [comment]
> LGTM. Is the readme regenerated (up-to-date) after the prop definition?

input-utils has no README to generate.

**Review by mohib0306 — APPROVED** [review]

File: packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx (hunk 1 above)
Comment (emmenko) [reply packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx:39]: Have you tried typing `useRef`?

File: packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx [hunk 4]
```diff
@@ -1,39 +1,65 @@
...
+      if (onHeightChange) {
+        onHeightChange(height, rowCount);
+      }
     },
     [ref, onHeightChange]
   );
   return (
     <TextareaAutosize
+      // @ts-ignore
       ref={ref}
```
Comment (emmenko) [inline packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx:49]: The `TextareaAutosize` exports `React.forwardRef`. What error do you get?

https://github.com/Andarist/react-textarea-autosize/blob/58c9807645dceb9034b733d9cb63aa66df2364b0/src/index.tsx#L101

File: packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx [hunk 5]
```diff
@@ -52,23 +78,4 @@ const MultilineInput = (props) => {
 
 MultilineInput.displayName = 'MultilineInput';
 
-MultilineInput.propTypes = {
-  name: PropTypes.string,
-  className: PropTypes.string,
-  autoComplete: PropTypes.string,
-  id: PropTypes.string,
-  value: PropTypes.string.isRequired,
-  onChange: requiredIf(PropTypes.func, (props) => !props.isReadOnly),
```
Comment (emmenko) [inline packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx:61]: We might want to add an warning/invariant to cover this case

**Review by emmenko — COMMENTED** [review]
Thanks, overall looks good. I left some comments in regards to the ignored type errors.

File: packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx (hunk 5 above)
Comment (adnasa) [reply packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx:61]: ah good, catch!

File: packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx (hunk 1 above)
Comment (adnasa) [reply packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx:39]: aah okay. I applied a type before but still got an error. This error is gone if I set the `scrollHeight` default value to `0`
Have a look at the diff. if we accept this, I'll proceed

<details>
 <summary>DIFF</summary>

```diff
diff --git a/packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx b/packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx
index 02da6c4f..75c3c537 100644
--- a/packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx
+++ b/packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx
@@ -20,39 +20,43 @@ export type TMultiLineInputProps = {
   name?: string;
   onBlur?: ChangeEventHandler;
   onChange?: ChangeEventHandler;
   onFocus?: ChangeEventHandler;
   placeholder?: string;
   value: string;
   isOpen: boolean;
   onHeightChange?: (height: number, rowCount: number) => void;
 };
 
+type TInputRef = {
+  scrollHeight: number;
+};
+
 const MultilineInput = (props: TMultiLineInputProps) => {
   const theme = useTheme();
   const { onHeightChange } = props;
-  const ref = React.useRef();
+
+  const ref = React.useRef<TInputRef>({ scrollHeight: 0 });
   const handleHeightChange = React.useCallback<
-    (height: number, meta?: TextareaHeightChangeMeta) => void
+    (height: number, meta: TextareaHeightChangeMeta) => void
   >(
-    (height: number, meta?: TextareaHeightChangeMeta) => {
-      // @ts-ignore
-      const rowCount = Math.floor(ref?.current?.scrollHeight / meta?.rowHeight);
+    (height: number, meta: TextareaHeightChangeMeta) => {
+      const rowCount = Math.floor(ref.current.scrollHeight / meta.rowHeight);
       if (onHeightChange) {
         onHeightChange(height, rowCount);
       }
     },
     [ref, onHeightChange]
   );
+
   return (
     <TextareaAutosize
-      // @ts-ignore
       ref={ref}
       name={props.name}
       onHeightChange={handleHeightChange}
       autoComplete={props.autoComplete}
       value={props.value}
       onChange={props.onChange}
       id={props.id}
       onBlur={props.onBlur}
       onFocus={props.onFocus}
       disabled={props.isDisabled}
```

</details>

File: packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx (hunk 4 above)
Comment (adnasa) [reply packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx:49]: it looks like a mismatch on the type of `useRef` and what is accepted as `ref` on the component

image

Comment (emmenko) [reply packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx:52]: Then use the correct type 😜

Comment (adnasa) [reply packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx:52]: which one then? 😄 

we are passing a `ref` into a component that extends `TextareaHTMLAttributes` with `IntrinsicAttributes`, none accepts `MutableRefObject`..
maybe I misunderstood something

Comment (adnasa) [reply packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx:52]: I just rubber ducked myself 🦆

Comment (adnasa) [reply packages/components/inputs/input-utils/src/multiline-input/multiline-input.tsx:51]: https://github.com/commercetools/ui-kit/pull/1867/commits/e9737e4067132362ecdca027375dfa1bd15f066f
